recipe-java-runtime/src/main/java/ca/derekcormier/recipe/Ingredient.java

```diff
@@ -14,6 +15,21 @@ public Ingredient(String type, String domain) {
         super(type, domain);
     }
 
+    protected <T extends Ingredient> T duplicate() {
+        Class clazz = this.getClass();
+        try {
+            Constructor<? extends Ingredient> constructor = clazz.getDeclaredConstructor(clazz);
+            Ingredient copy = constructor.newInstance(this);
+            this.getProperties().forEach((key, value) -> {
+                copy.setProperty(key, value);
+            });
+            return (T)copy;
+        }
+        catch (Exception e) {
+            throw new RuntimeException("could not call empty constructor on ingredient " + clazz.getSimpleName(), e);
```

Comment:
"empty constructor"? But it's not empty. Maybe refer to it as the "copy constructor"?

Reply:
oops